File: typescript/core/src/context.ts
/**
 * Project context loader — AGENTS.md (or its fallbacks) plus resolved file
 * references. The TypeScript port of the Rust reference
 * `smooth-operator-core::context` (pearl th-5002c4).
 *
 * Smooth previously only read AGENTS.md. Many projects don't have one but DO
 * have CLAUDE.md or a SMOOTH.md or .smooth/CONTEXT.md. User-level facts also
 * belong in the prompt, so we walk a preference order and STACK user-level +
 * project-level context.
 *
 * Preference order (first hit per layer; layers stack):
 *
 * - USER layer (read once, prepended):
 *   `~/.smooth/CONTEXT.md` → `~/.smooth/AGENTS.md` → `~/.smooth/CLAUDE.md`
 *
 * - PROJECT layer (walk up from `workingDir`, first hit wins):
 *   `<dir>/.smooth/CONTEXT.md` → `<dir>/SMOOTH.md` → `<dir>/AGENTS.md` → `<dir>/CLAUDE.md`
 *
 * AGENTS.md / SMOOTH.md can carry file references in a `## File References`
 * section:
 *
 * ```markdown
 * ## File References
 * - [CLAUDE.md](CLAUDE.md) — full file
 * - [Section name](CLAUDE.md#6-pearl-tracking) — specific section
 * ```
 *
 * Those are resolved against the file's directory and appended inline. The
 * combined string is what the host injects into the agent's system prompt —
 * like the Rust reference, this is a standalone loader, NOT a hook inside the
 * agent loop.
 */

import { readFileSync, statSync } from 'node:fs';
import { homedir } from 'node:os';
import { dirname, join } from 'node:path';

/** A parsed file reference from the `## File References` section. */
export interface FileRef {
    /** Display label from the markdown link text. */
    label: string;
    /** Relative file path (without fragment). */
    path: string;
    /** Optional `#fragment` pointing to a heading. */
    fragment?: string;
    /** Optional description after the ` — `. */
    description?: string;
}

const USER_CONTEXT_CANDIDATES = ['.smooth/CONTEXT.md', '.smooth/AGENTS.md', '.smooth/CLAUDE.md'];
const PROJECT_CONTEXT_CANDIDATES = ['.smooth/CONTEXT.md', 'SMOOTH.md', 'AGENTS.md', 'CLAUDE.md'];

/**
 * Load the combined project + user context, user-level prepended, with file
 * references in any AGENTS.md / SMOOTH.md resolved inline.
 *
 * `undefined` only when NEITHER layer found anything — so a workspace with a
 * bare CLAUDE.md and no user-level file still loads context.
 */
export function loadProjectContext(workingDir: string): string | undefined {
    const user = loadUserContext();
    const project = loadLayeredProjectContext(workingDir);

    if (user === undefined && project === undefined) return undefined;
    if (project === undefined) return `## User context (~/.smooth)\n\n${user}`;
    if (user === undefined) return project;
    return `## User context (~/.smooth)\n\n${user}\n\n---\n\n${project}`;
}

/** First non-blank hit from the user-level preference list. */
function loadUserContext(): string | undefined {
    const home = homedir();
    if (!home) return undefined;
    for (const candidate of USER_CONTEXT_CANDIDATES) {
        const raw = readFileOrUndefined(join(home, candidate));
        if (raw !== undefined && raw.trim() !== '') return raw;
    }
    return undefined;
}

/** The project context file with its file references resolved. */
function loadLayeredProjectContext(workingDir: string): string | undefined {
    const contextPath = findProjectContextFile(workingDir);
    if (contextPath === undefined) return undefined;
    const raw = readFileOrUndefined(contextPath);
    if (raw === undefined) return undefined;

    const refs = parseFileReferences(raw);
    if (refs.length === 0) return raw;

    const resolved = resolveReferences(dirname(contextPath), refs);
    if (resolved.length === 0) return raw;

    let output = raw;
    output += '\n---\n\n## Resolved File References\n\n';
    for (const { ref, content } of resolved) {
        output += ref.description === undefined ? `### ${ref.label}\n` : `### ${ref.label} — ${ref.description}\n`;
        output += '\n```\n';
        output += content;
        if (!content.endsWith('\n')) output += '\n';
        output += '```\n\n';
    }
    return output;
}

/**
 * Walk up from `startDir` looking for a project context file. Preference order
 * at each level: `.smooth/CONTEXT.md` → `SMOOTH.md` → `AGENTS.md` →
 * `CLAUDE.md`. First hit wins per directory, then keep walking up.
 */
export function findProjectContextFile(startDir: string): string | undefined {
    let dir = startDir;
    for (;;) {
        for (const candidate of PROJECT_CONTEXT_CANDIDATES) {
            const path = join(dir, candidate);
            try {
                if (statSync(path).isFile()) return path;
            } catch {
                // Not there — try the next candidate.
            }
        }
        const parent = dirname(dir);
        if (parent === dir) return undefined;
        dir = parent;
    }
}

/**
 * Parse the `## File References` section out of AGENTS.md content. Expects
 * markdown list items like `- [Label](path.md#fragment) — description`.
 */
export function parseFileReferences(content: string): FileRef[] {
    const refs: FileRef[] = [];
    let inSection = false;

    for (const line of splitLines(content)) {
        const trimmed = line.trim();

        // Detect the file references section.
        if (trimmed.startsWith('## ') || trimmed.startsWith('# ')) {
            inSection = trimmed.toLowerCase().includes('file reference');
            continue;
        }
        if (!inSection) continue;

        const ref = parseLinkLine(trimmed);
        if (ref !== undefined) refs.push(ref);
    }
    return refs;
}

/** Parse a single markdown list-item link line. */
export function parseLinkLine(line: string): FileRef | undefined {
    // Strip leading `- ` or `* `.
    let rest: string;
    if (line.startsWith('- ')) rest = line.slice(2);
    else if (line.startsWith('* ')) rest = line.slice(2);
    else return undefined;

    // Match [label](target).
    const openBracket = rest.indexOf('[');
    if (openBracket < 0) return undefined;
    const closeBracket = rest.indexOf(']', openBracket);
    if (closeBracket < 0) return undefined;
    const label = rest.slice(openBracket + 1, closeBracket);

    const after = rest.slice(closeBracket + 1);
    const openParen = after.indexOf('(');
    if (openParen < 0) return undefined;
    const closeParen = after.indexOf(')', openParen);
    if (closeParen < 0) return undefined;
    const target = after.slice(openParen + 1, closeParen);

    // Split path and fragment.
    const hash = target.indexOf('#');
    const path = hash < 0 ? target : target.slice(0, hash);
    const fragment = hash < 0 ? undefined : target.slice(hash + 1);

    // Description after ` — ` / ` - ` / ` -- `.
    const afterLink = after.slice(closeParen + 1);
    let description: string | undefined;
    for (const sep of [' — ', ' - ', ' -- ']) {
        if (afterLink.startsWith(sep)) {
            const d = afterLink.slice(sep.length).trim();
            if (d !== '') description = d;
            break;
        }
    }

    if (path === '' && fragment === undefined) return undefined;
    return { label, path, fragment, description };
}

/** Resolve file references against a base directory, skipping unreadable ones. */
function resolveReferences(baseDir: string, refs: FileRef[]): { ref: FileRef; content: string }[] {
    const results: { ref: FileRef; content: string }[] = [];
    for (const ref of refs) {
        const raw = readFileOrUndefined(join(baseDir, ref.path));
        if (raw === undefined) continue; // Skip unreadable files.
        const content = ref.fragment === undefined ? raw : extractSection(raw, ref.fragment);
        if (content.trim() !== '') results.push({ ref, content });
    }
    return results;
}

/**
 * Extract a markdown section by heading fragment. The fragment is matched
 * against GitHub-style heading anchors; the section runs until the next
 * heading at the same or a higher level.
 */
export function extractSection(content: string, fragment: string): string {
    const target = headingToAnchor(fragment);
    const lines = splitLines(content);
    let start = -1;
    let startLevel = 0;

    for (let i = 0; i < lines.length; i++) {
        const heading = parseHeading(lines[i]!);
        if (heading === undefined) continue;
        const anchor = headingToAnchor(heading.text);
        if (anchor === target || anchor.includes(target) || target.includes(anchor)) {
            start = i;
            startLevel = heading.level;
            continue;
        }
        // Started capturing and hit a same-or-higher-level heading: stop.
        if (start >= 0 && heading.level <= startLevel) return lines.slice(start, i).join('\n');
    }

    return start >= 0 ? lines.slice(start).join('\n') : '';
}

/** Parse a markdown heading line into its level and text. */
function parseHeading(line: string): { level: number; text: string } | undefined {
    const trimmed = line.trim();
    if (!trimmed.startsWith('#')) return undefined;
    const level = trimmed.length - trimmed.replace(/^#+/, '').length;
    const text = trimmed.slice(level).trim();
    if (text === '') return undefined;
    return { level, text };
}

/** Convert heading text to a GitHub-style anchor. */
export function headingToAnchor(text: string): string {
    let out = '';
    for (const ch of text.toLowerCase()) {
        if (/[\p{L}\p{N}\-_]/u.test(ch)) out += ch;
        else if (ch === ' ') out += '-';
        // Other characters are dropped.
    }
    // Single non-overlapping pass, matching Rust's `str::replace`.
    return out.replaceAll('--', '-');
}

/** `undefined` instead of throwing, for any unreadable path. */
function readFileOrUndefined(path: string): string | undefined {
    try {
        return readFileSync(path, 'utf8');
    } catch {
        return undefined;
    }
}

/**
 * Mirrors Rust's `str::lines`: split on \n, drop a trailing \r on each line,
 * and emit no trailing empty line for a final newline.
 */
function splitLines(s: string): string[] {
    if (s === '') return [];
    const body = s.endsWith('\n') ? s.slice(0, -1) : s;
    return body.split('\n').map((line) => (line.endsWith('\r') ? line.slice(0, -1) : line));
}

File: python/core/src/smooth_operator_core/context.py
"""Project context loader — AGENTS.md (or its fallbacks) plus resolved file references.

The Python port of the Rust reference engine's ``context.rs`` (pearl th-5002c4).

Smooth previously only read AGENTS.md. Many projects don't have one but DO have
CLAUDE.md or a SMOOTH.md or ``.smooth/CONTEXT.md``. User-level facts also belong in the
prompt, so we walk a preference order and **stack** user-level + project-level context.

Preference order (first hit per layer; layers stack):

- USER layer (read once, prepended):
  ``~/.smooth/CONTEXT.md`` → ``~/.smooth/AGENTS.md`` → ``~/.smooth/CLAUDE.md``

- PROJECT layer (walk up from ``working_dir``, first hit wins):
  ``<dir>/.smooth/CONTEXT.md`` → ``<dir>/SMOOTH.md`` → ``<dir>/AGENTS.md`` → ``<dir>/CLAUDE.md``

AGENTS.md / SMOOTH.md can carry file references in a ``## File References`` section::

    ## File References
    - [CLAUDE.md](CLAUDE.md) — full file
    - [Section name](CLAUDE.md#6-pearl-tracking) — specific section

Those are resolved against the file's directory and appended inline. The combined
string is what the host injects into the agent's system prompt — like the Rust
reference, this is a standalone loader, **not** a hook inside the agent loop.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

USER_CONTEXT_CANDIDATES = (".smooth/CONTEXT.md", ".smooth/AGENTS.md", ".smooth/CLAUDE.md")
PROJECT_CONTEXT_CANDIDATES = (".smooth/CONTEXT.md", "SMOOTH.md", "AGENTS.md", "CLAUDE.md")


@dataclass(frozen=True)
class FileRef:
    """A parsed file reference from the ``## File References`` section."""

    label: str
    """Display label from the markdown link text."""
    path: str
    """Relative file path (without fragment)."""
    fragment: str | None = None
    """Optional ``#fragment`` pointing to a heading."""
    description: str | None = None
    """Optional description after the ``` — ```."""


def load_project_context(working_dir: str | Path) -> str | None:
    """Load the combined project + user context, user-level prepended.

    File references in any AGENTS.md / SMOOTH.md are resolved inline. Returns ``None``
    only when NEITHER layer found anything — so a workspace with a bare CLAUDE.md and
    no user-level file still loads context.
    """
    user = _load_user_context()
    project = _load_layered_project_context(Path(working_dir))

    if user is None and project is None:
        return None
    if project is None:
        return f"## User context (~/.smooth)\n\n{user}"
    if user is None:
        return project
    return f"## User context (~/.smooth)\n\n{user}\n\n---\n\n{project}"


def _load_user_context() -> str | None:
    """First non-blank hit from the user-level preference list."""
    try:
        home = Path.home()
    except (RuntimeError, OSError):  # No home dir (minimal CI / broken containers).
        return None
    for candidate in USER_CONTEXT_CANDIDATES:
        raw = _read_or_none(home / candidate)
        if raw is not None and raw.strip():
            return raw
    return None


def _load_layered_project_context(working_dir: Path) -> str | None:
    """The project context file with its file references resolved."""
    context_path = _find_project_context_file(working_dir)
    if context_path is None:
        return None
    raw = _read_or_none(context_path)
    if raw is None:
        return None

    refs = parse_file_references(raw)
    if not refs:
        return raw

    resolved = _resolve_references(context_path.parent, refs)
    if not resolved:
        return raw

    parts = [raw, "\n---\n\n## Resolved File References\n\n"]
    for ref, content in resolved:
        parts.append(f"### {ref.label}\n" if ref.description is None else f"### {ref.label} — {ref.description}\n")
        parts.append("\n```\n")
        parts.append(content)
        if not content.endswith("\n"):
            parts.append("\n")
        parts.append("```\n\n")
    return "".join(parts)


def _find_project_context_file(start_dir: Path) -> Path | None:
    """Walk up from ``start_dir``; first candidate hit per directory wins."""
    directory = start_dir
    while True:
        for candidate in PROJECT_CONTEXT_CANDIDATES:
            path = directory / candidate
            if path.is_file():
                return path
        parent = directory.parent
        if parent == directory:
            return None
        directory = parent


def parse_file_references(content: str) -> list[FileRef]:
    """Parse the ``## File References`` section out of AGENTS.md content.

    Expects markdown list items like ``- [Label](path.md#fragment) — description``.
    """
    refs: list[FileRef] = []
    in_section = False

    for line in _split_lines(content):
        trimmed = line.strip()

        # Detect the file references section.
        if trimmed.startswith("## ") or trimmed.startswith("# "):
            in_section = "file reference" in trimmed.lower()
            continue
        if not in_section:
            continue

        ref = _parse_link_line(trimmed)
        if ref is not None:
            refs.append(ref)
    return refs


def _parse_link_line(line: str) -> FileRef | None:
    """Parse a single markdown list-item link line."""
    # Strip leading `- ` or `* `.
    if line.startswith("- ") or line.startswith("* "):
        rest = line[2:]
    else:
        return None

    # Match [label](target).
    open_bracket = rest.find("[")
    if open_bracket < 0:
        return None
    close_bracket = rest.find("]", open_bracket)
    if close_bracket < 0:
        return None
    label = rest[open_bracket + 1 : close_bracket]

    after = rest[close_bracket + 1 :]
    open_paren = after.find("(")
    if open_paren < 0:
        return None
    close_paren = after.find(")", open_paren)
    if close_paren < 0:
        return None
    target = after[open_paren + 1 : close_paren]

    # Split path and fragment.
    hash_pos = target.find("#")
    path = target if hash_pos < 0 else target[:hash_pos]
    fragment = None if hash_pos < 0 else target[hash_pos + 1 :]

    # Description after ` — ` / ` - ` / ` -- `.
    after_link = after[close_paren + 1 :]
    description = None
    for sep in (" — ", " - ", " -- "):
        if after_link.startswith(sep):
            candidate = after_link[len(sep) :].strip()
            if candidate:
                description = candidate
            break

    if not path and fragment is None:
        return None
    return FileRef(label=label, path=path, fragment=fragment, description=description)


def _resolve_references(base_dir: Path, refs: list[FileRef]) -> list[tuple[FileRef, str]]:
    """Resolve references against ``base_dir``, skipping unreadable or empty ones."""
    results: list[tuple[FileRef, str]] = []
    for ref in refs:
        raw = _read_or_none(base_dir / ref.path)
        if raw is None:
            continue  # Skip unreadable files.
        content = raw if ref.fragment is None else _extract_section(raw, ref.fragment)
        if content.strip():
            results.append((ref, content))
    return results


def _extract_section(content: str, fragment: str) -> str:
    """Extract a markdown section by heading fragment.

    The fragment is matched against GitHub-style heading anchors; the section runs
    until the next heading at the same or a higher level.
    """
    target = _heading_to_anchor(fragment)
    lines = _split_lines(content)
    start: int | None = None
    start_level = 0

    for i, line in enumerate(lines):
        heading = _parse_heading(line)
        if heading is None:
            continue
        level, text = heading
        anchor = _heading_to_anchor(text)
        if anchor == target or target in anchor or anchor in target:
            start = i
            start_level = level
            continue
        # Started capturing and hit a same-or-higher-level heading: stop.
        if start is not None and level <= start_level:
            return "\n".join(lines[start:i])

    return "\n".join(lines[start:]) if start is not None else ""


def _parse_heading(line: str) -> tuple[int, str] | None:
    """Parse a markdown heading line into ``(level, text)``."""
    trimmed = line.strip()
    if not trimmed.startswith("#"):
        return None
    level = len(trimmed) - len(trimmed.lstrip("#"))
    text = trimmed[level:].strip()
    if not text:
        return None
    return level, text


def _heading_to_anchor(text: str) -> str:
    """Convert heading text to a GitHub-style anchor."""
    out: list[str] = []
    for ch in text.lower():
        if ch.isalnum() or ch in "-_":
            out.append(ch)
        elif ch == " ":
            out.append("-")
        # Other characters are dropped.
    # Single non-overlapping pass, matching Rust's `str::replace`.
    return "".join(out).replace("--", "-")


def _read_or_none(path: Path) -> str | None:
    """``None`` instead of raising, for any unreadable path."""
    try:
        return path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return None


def _split_lines(s: str) -> list[str]:
    """Mirror Rust's ``str::lines``: split on ``\\n`` only, strip a trailing ``\\r``.

    ``str.splitlines`` would also break on ``\\v``, ``\\f`` and Unicode separators,
    which the Rust reference does not.
    """
    if not s:
        return []
    body = s[:-1] if s.endswith("\n") else s
    return [line[:-1] if line.endswith("\r") else line for line in body.split("\n")]
